Clarify named tracers (#5)
dynatrace-oss-contrib/oteps: text/0016-named-tracers.md

**Status:** `approved`

> _Creating Tracers and Meters using a Registry mechanism and naming those Tracers / Meters in accordance with the library that provides the instrumentation for those components._

Reviewer: I'm not entirely happy with both factory and registry (although registry is definitely better). Also I think we should get rid of the verb "naming" here, since we said we want to abolish the "named tracers" terminology (how about we also rename the file, btw?). Also, what is "those" referring to in that sentence?  How about:

```suggestion
_Associate potentially different Tracers and Meters along with meta-information with each library that implements instrumentation by parametrizing the APIs that create Tracers and Meters._
```

Reviewer: I think this change is still fine, even if we don't want to get rid of "named" terminology.

Author: What would you think of:

>Associate Tracers and Meters with the name and version of the library which reports telemetry data by parameterizing the API which the library uses to acquire the Tracer or Meter.

## Motivation

The mechanism of "Named Tracers and Meters" proposed here is motivated by the following scenarios:

> **Faulty or expensive instrumentation**

Reviewer: ```suggestion
### Faulty or expensive instrumentation
```

And so on.

For an operator of an application using OpenTelemetry, there is currently no way to influence the amount of data produced by reporting libraries. Reporting libraries can easily "spam" backend systems, deliver bogus data, or -- in the worst case -- crash or slow down applications. These problems might even occur suddenly in production environments because of external factors such as increasing load or unexpected input data.

**Reporting library identification**

If a reporting library hasn't implemented [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md) correctly or those conventions change over time, it's currently hard to interpret and sanitize data produced by it selectively. The produced Spans or Metrics cannot later be associated with the library which reported them, either in the processing pipeline or the backend.

> **Conflicting instrumentation libraries**

Reviewer 2: I believe this is a very important point and I'd add at least also this example (please feel free to re-phrase :) ):

Conflict of "pre-instrumented library" and "plugin instrumentation (loaded as a separate dependency)". A user may want to chose to use "plugin instrumentation" and disable built-in instrumentation.

Author: That was the original intention of this section, so obviously it wasn't clear enough. Is this better?

>It is the eventual goal of OpenTelemetry that library vendors implement the OpenTelemetry API, obviating the need to auto-instrument their library. An operator should be able to disable the telemetry that is built into some database driver or other library and provide their own integration if the built-in telemetry is lacking in some way. This should be possible even if the developer of that database driver has not provided a configuration to disable telemetry.

It is the eventual goal of OpenTelemetry that library vendors implement the OpenTelemetry API, obviating the need to auto-instrument their library. An operator should be able to disable built-in instrumentation even if the library does not provide an explicit configuration to disable instrumentation.

> This proposal attempts to solve the stated problems by introducing the concept of:

Reviewer: There needs to be a new heading here, otherwise it looks like these paragraphs belong to `Conflicting instrumentation libraries`

### Reporting library identification

If a reporting library hasn't implemented [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md) correctly or those conventions change over time, it's currently hard to interpret and sanitize data produced by it selectively. The produced Spans or Metrics cannot later be associated with the library which reported them, either in the processing pipeline or the backend.

### Disable instrumentation of pre-instrumented libraries

It is the eventual goal of OpenTelemetry that library vendors implement the OpenTelemetry API, obviating the need to auto-instrument their library. An operator should be able to disable the telemetry that is built into some database driver or other library and provide their own integration if the built-in telemetry is lacking in some way. This should be possible even if the developer of that database driver has not provided a configuration to disable telemetry.

> ## Solution

Reviewer: I think when conforming to the [OTEP template](https://github.com/open-telemetry/oteps/blob/master/0000-template.md), this would be the "Explanation" but I wonder if we can find a good sub-heading for our current explanation, and whether our current "Internal details" vs "Explanation" division makes sense. This now sounds like a can of worms, so "Solution" should actually be fine. 😄

This proposal attempts to solve the stated problems by introducing the concept of:
 * _Named Tracers and Meters_ which are associated with the **name** (e.g. _"io.opentelemetry.contrib.mongodb"_) and **version** (e.g._"semver:1.0.0"_) of the library which acquired them.

>  * A `TracerRegistry` / `MeterRegistry` as the only means of creating a Tracer or Meter.

Reviewer: ```suggestion
 * A `TracerRegistry` / `MeterRegistry` as the only means of acquiring a Tracer or Meter.
```

I think "acquire" (or alternatively "obtain" or "access") sounds better in combination with a Registry than "create". Not entirely sure though.

> Based on such an identifier, a Registry could provide a no-op Tracer or Meter to specific instrumentation libraries, or a Sampler could be implemented that discards Spans or Metrics from certain libraries. Also, by providing custom Exporters, Span or Metric data could be sanitized before it gets processed in a back-end system. However, this is beyond the scope of this proposal, which only provides the fundamental mechanisms.

Reviewer: ```suggestion
Based on that information, a Registry could provide a no-op Tracer or Meter to specific instrumentation libraries, or a Sampler could be implemented that discards Spans or Metrics from certain libraries. Also, by providing custom Exporters, Span or Metric data could be sanitized before it gets processed in a back-end system. However, this is beyond the scope of this proposal, which only provides the fundamental mechanisms.
```

> The mechanism of "Named Tracers and Meters" proposed here is motivated by following scenarios:

Reviewer: ```suggestion
The mechanism of "Named Tracers and Meters" proposed here is motivated by the following scenarios:
```

> **Expensive instrumentation**

Reviewer 3: The paragraph below not only covers excessive instrumentation but also buggy ones which deliver incorrect data. What about "Faulty or excessive instrumentation"?

> For an operator of OpenTelemetry, there is currently no possibility of influencing the amount of the data produced by reporting libraries. Reporting libraries can easily "spam" backend systems, deliver bogus data, or - in the worst case - crash or slow down applications. These problems might even occur suddenly in production environments caused by external factors such as increasing load or unexpected input data.

Reviewer 3: ```suggestion
For an operator of OpenTelemetry, there is currently no possibility of influencing the amount of data produced by reporting libraries. Reporting libraries can easily "spam" backend systems, deliver bogus data, or - in the worst case - crash or slow down applications. These problems might even occur suddenly in production environments caused by external factors such as increasing load or unexpected input data.
```

> For an operator of OpenTelemetry, there is currently no possibility of influencing the amount of the data produced by reporting libraries. Reporting libraries can easily "spam" backend systems, deliver bogus data, or - in the worst case - crash or slow down applications. These problems might even occur suddenly in production environments caused by external factors such as increasing load or unexpected input data.

Reviewer: ```suggestion
For an operator of an application using OpenTelemetry, there is currently no way to influence the amount of data produced by reporting libraries. Reporting libraries can easily "spam" backend systems, deliver bogus data, or -- in the worst case -- crash or slow down applications. These problems might even occur suddenly in production environments because of external factors such as increasing load or unexpected input data.
```

**Reporting library identification**

> If a reporting library hasn't implemented [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md) correctly or those conventions change over time, it's currently hard to interpret and sanitize these data selectively. The produced Spans or Metrics cannot be associated with the library which reported them later.

Reviewer 3: ```suggestion
If a reporting library hasn't implemented [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md) correctly or those conventions change over time, it's currently hard to interpret and sanitize these data selectively. The produced Spans or Metrics cannot be associated with the library which reported them - neither in the processing pipeline nor on the backend.
```

> If a reporting library hasn't implemented [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md) correctly or those conventions change over time, it's currently hard to interpret and sanitize these data selectively. The produced Spans or Metrics cannot be associated with the library which reported them later.

Reviewer: ```suggestion
If a reporting library hasn't implemented [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md) correctly or those conventions change over time, it's currently hard to interpret and sanitize data produced by it selectively. The produced Spans or Metrics cannot be associated with the library which reported them later.
```

**Conflicting instrumentation libraries**

> It is the eventual goal of OpenTelemetry that library vendors implement the OpenTelemetry API, obviating the need to auto-instrument their library. An operator should be able to disable built-in instrumentation even if the library does not provide an explicit configuration to disable instrumentation.

Reviewer: This use case seems to fall under "Expensive Instrumenation" (or generally "excessive instrumentation"). The conflict would occur if in addition to the library's built-in instrumentation, there is also an auto-instrumentation or an integration that is hard-coded as activated. Only then, does it make sense to use reporter-information to disable/discard the spans produced by one of the conflicting reporters.

Author: Not true. I may want to disable the OpenTelemetry integration that is built into some database driver in order to provide my own integration if the built-in version is lacking in some way.

Reviewer: OK, that makes sense. But then I would add a sentence indicating that here, because as it stands now, this seems to describe only one side of the conflict.

Author: To me, the paragraph _only_ has the use case of disabling the built-in instrumentation of some library. Would this be more clear?

>It is the eventual goal of OpenTelemetry that library vendors implement the OpenTelemetry API, obviating the need to auto-instrument their library. An operator should be able to disable the telemetry that is built into some database driver or other library and provide their own integration if the built-in telemetry is lacking in some way. This should be possible even if the developer of that database driver has not provided a configuration to disable telemetry.

@Oberon00 WDYT?

This proposal attempts to solve the stated problems by introducing the concept of:

>  * _Named Tracers and Meters_ identified via a **name** (e.g. _"io.opentelemetry.contrib.mongodb"_) and a **version** (e.g._"semver:1.0.0"_) which is associated with the Tracer / Meter and the Spans / Metrics it produces.

Reviewer: Do we really want tracers to be *identifieable* by this information? Or is it enough to have the reporter information associated with them?

Author: What about: 

> * _Named Tracers and Meters_ which are associated with the **name** (e.g. _"io.opentelemetry.contrib.mongodb"_) and **version** (e.g._"semver:1.0.0"_) of the library which acquired them.

>  * A `TracerRegistry` / `MeterRegistry` as the only means of creating a Tracer or Meter which stores references to all `Tracers` and `Meters` it has created.

Reviewer: Why do we need to prescribe storing of references? I personally don't think this is needed or even useful.

Author: This was just so the registry could return the same tracer if a library acquires it more than once. Would you have it create a new tracer every time getTracer is called?

Reviewer: That is one possibility. In Python I did in fact implement it that way, since the tracers consist of just two properties, the Info object and a back-reference to the tracer registry. I don't think storing references is wrong, I just don't think we should demand that. But it may actually make sense to add that sentence back but reformulate it to something like "Note that this could be implemented by ..."

Reviewer 3: I think Christian's intent was to not have the spec define such implementation details.

Author: I just removed it. We can have a line like that in the spec if we want.

> Based on such an identifier, a Registry could provide a no-op Tracer or Meter to specific instrumentation libraries, or a Sampler could be implemented that discards Spans or Metrics from certain libraries. Also, by providing custom Exporters, Span or Metric data could be sanitized before it gets processed in a back-end system. However, this is beyond the scope of this proposal, which only provides the fundamental mechanisms.

Reviewer: -> https://github.com/dynatrace-oss-contrib/oteps/pull/5/files#r363729699

If a library (or application) has instrumentation built-in, it is both the instrumenting and instrumented library and should pass its own name here. In all other cases (and to distinguish them from that case), the distinction between instrumenting and instrumented library is very important. For example, if an HTTP library `com.example.http` is instrumented by either `io.opentelemetry.contrib.examplehttp`, then it is important that the Tracer is not named `com.example.http`, but `io.opentelemetry.contrib.examplehttp` after the actual instrumentation library.

If no name (null or empty string) is specified, following the suggestions in ["error handling proposal"](https://github.com/open-telemetry/opentelemetry-specification/pull/153), a "smart default" will be applied and a default Tracer / Meter implementation is returned.

### Examples (of Tracer and Meter names)

Since Tracer and Meter names describe the libraries which use those Tracers and Meters, their names should be defined in a way that makes them as unique as possible.

> The name of the Tracer / Meter should represent the identity of the library, class or package that provides the instrumentation.

Reviewer: When using the "reporter info" terminology instead of "named XY", this would become something like:

```suggestion
The reporter of a span is the library, class or package that provides instrumentation (usually that is the library that uses the OpenTelemetry APIs).
```

examples:
* `@opentelemetry/plugin-http`
* `io.opentelemetry.redis`
* `redis-client` (in this case, `redis-client` has instrumented itself with the OpenTelemetry API)

#### Tracer name / Meter name
When it is acquired from the tracer/meter registry, the tracer/meter is assigned a name and a version which is the name and version of the reporting library that acquired the tracer. In cases where a library has instrumented itself using the OpenTelemetry API, they may be the same.

example: If the `http` library is being instrumented by a library with the name `io.opentelemetry.contrib.http`, then the tracer name is also `io.opentelemetry.contrib.http`. If that same `http` library has built-in instrumentation through use of the OpenTelemetry API, then the tracer name would be `http`.

> ### Disambiguation with related terms

Author: I'm not sure this section should be in here. WDYT?

Reviewer: I think having this section here is absolutely fine.

#### Meter namespace
This is included here because it is often confused with the meter name.

> This describes the component which is being instrumented. For instance, `latency` could be collected from a number of different things (http latency, disk latency, event loop latency) and a namespace ties the `latency` metric to the entity being metered.

Reviewer 2: "This describes the component which is being instrumented"
->
"A meter namespace describes the component which is being instrumented"

Also: Is there another page where meter namespace is described? Would it make sense to link it?

Author: I was looking for other references to namespace and the spec actually seems to disagree with us in it's current state:

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics-user.md#metric-names

>Metrics names belong to a namespace by virtue of a "Named" Meter instance. A "Named" Meter refers to the requirement that every Meter instance must have an associated component label, determined statically in the code. The component label value of the associated Meter serves as its namespace, allowing the same metric name to be used in multiple libraries of code, unambiguously, within the same application.

Author: There is also nowhere else in oteps or spec with namespaces. This might be part of why people are so confused.

Reviewer 2: Oh, good catch. That spec part part we should also try to fix.
